File: quantsolve-server/src/engine/engineRunner.js
// engine/engineRunner.js
// Orchestrates the full QuantSolve pipeline:
//   Input → Lexer → Parser → Normalizer → Solver (worker thread) → Formatter
//
// Solver runs in a worker_threads Worker so it never blocks the main event loop.
// Response is PAGINATED — solver finds all solutions, sends only the requested page.

const { Worker } = require("worker_threads");
const path = require("path");

const { lexer } = require("./lexer");
const Parser = require("./parser");
const normalizeEquation = require("./normalizer");
const formatResults = require("./resultFormatter");
const handleError = require("./errorHandler");
const { EngineError, ErrorCode } = require("./errors");
const { trySolvePolynomial } = require("./polynomialSolver");

const SOLVE_TIMEOUT_MS = 8000;

function runSolverInWorker(coeffs, target, constraints, options) {
    return new Promise((resolve, reject) => {
        const worker = new Worker(path.join(__dirname, "solverWorker.js"), {
            workerData: { coeffs, target, constraints, options },
        });

        const timer = setTimeout(() => {
            worker.terminate();
            reject(new EngineError(ErrorCode.UNBOUNDED_SEARCH));
        }, SOLVE_TIMEOUT_MS);

        worker.once("message", (result) => {
            clearTimeout(timer);
            if (result.success) {
                resolve(result);
            } else {
                reject(
                    Object.assign(new Error(result.error), {
                        isEngineError: true,
                        code: result.code,
                    }),
                );
            }
        });

        worker.once("error", (err) => {
            clearTimeout(timer);
            reject(err);
        });

        worker.once("exit", (code) => {
            clearTimeout(timer);
            if (code !== 0) {
                reject(
                    new EngineError(
                        ErrorCode.INTERNAL_ERROR,
                        `Worker exited with code ${code}`,
                    ),
                );
            }
        });
    });
}

async function runEngine(input, userConstraints = {}, options = {}) {
    try {
        if (typeof input !== "string" || !input.trim()) {
            throw new EngineError(ErrorCode.EMPTY_INPUT);
        }

        const cleanedInput = input.replace(/\s+/g, " ").trim();

        // Phase 1: Lex
        const tokens = lexer(cleanedInput);

        // Phase 2: Parse
        const parser = new Parser(tokens);
        const { left, right } = parser.parseEquation();

        // Solver options are shared by linear and polynomial paths.
        const workerOptions = {
            page: options.page || 1,
            pageSize: options.pageSize || 50,
        };

        // Optional polynomial path (currently single-variable integer polynomial support).
        const polyResult = trySolvePolynomial(
            left,
            right,
            userConstraints,
            workerOptions,
        );

        if (polyResult) {
            const formatted = formatResults(polyResult.solutions, {
                totalFound: polyResult.totalFound,
                variableOrder: polyResult.variableOrder,
                page: polyResult.page,
                pageSize: polyResult.pageSize,
                totalPages: polyResult.totalPages,
                hasMore: polyResult.hasMore,
            });

            function treeDepth(node) {
                if (!node) return 0;
                if (node.type === "Number" || node.type === "Variable") return 1;
                if (node.type === "UnaryOp") return 1 + treeDepth(node.operand);
                return 1 + Math.max(treeDepth(node.left), treeDepth(node.right));
            }

            return {
                success: true,
                input: cleanedInput,
                coeffs: {},
                target: 0,
                variableOrder: polyResult.variableOrder,
                polynomial: polyResult.polynomial,
                ast: { left, right },
                totalFound: polyResult.totalFound,
                page: polyResult.page,
                pageSize: polyResult.pageSize,
                totalPages: polyResult.totalPages,
                hasMore: polyResult.hasMore,
                solutions: polyResult.solutions,
                formattedResult: formatted,
                warnings: formatted.warnings,
                meta: {
                    variableCount: polyResult.variableOrder.length,
                    constraintCount: Object.keys(userConstraints).length,
                    astDepth: treeDepth(left),
                    solverType: "polynomial",
                    polynomialDegree: polyResult.polynomial.degree,
                    searchBounds: polyResult.searchBounds,
                },
            };
        }

        // Phase 3: Normalize (linear path)
        const { coeffs, target } = normalizeEquation(left, right);
        const variableOrder = Object.keys(coeffs);

        // Phase 4: Solve (worker thread)
        // Pass page + pageSize so solver returns only the requested slice.
        // totalFound in the response is always the TRUE full count.
        const solveResult = await runSolverInWorker(
            coeffs,
            target,
            userConstraints,
            workerOptions,
        );

        // Phase 5: Format
        const formatted = formatResults(solveResult.solutions, {
            totalFound: solveResult.totalFound,
            capped: false,
            variableOrder,
            page: solveResult.page,
            pageSize: solveResult.pageSize,
            totalPages: solveResult.totalPages,
            hasMore: solveResult.hasMore,
        });

        function treeDepth(node) {
            if (!node) return 0;
            if (node.type === "Number" || node.type === "Variable") return 1;
            if (node.type === "UnaryOp") return 1 + treeDepth(node.operand);
            return 1 + Math.max(treeDepth(node.left), treeDepth(node.right));
        }

        return {
            success: true,
            input: cleanedInput,
            coeffs,
            target,
            variableOrder,
            ast: { left, right },
            // Pagination metadata
            totalFound: solveResult.totalFound, // true total e.g. 352,800
            page: solveResult.page,
            pageSize: solveResult.pageSize,
            totalPages: solveResult.totalPages,
            hasMore: solveResult.hasMore,
            // This page's solutions
            solutions: solveResult.solutions,
            formattedResult: formatted,
            warnings: formatted.warnings,
            meta: {
                variableCount: Object.keys(coeffs).length,
                constraintCount: Object.keys(userConstraints).length,
                astDepth: treeDepth(left),
                solverType: "linear",
            },
        };
    } catch (err) {
        const errorPayload = handleError(err);
        return {
            ...errorPayload,
            input,
            solutions: [],
            formattedResult: { rows: [], count: 0, warnings: [] },
            warnings: [],
            meta: {},
        };
    }
}

module.exports = { runEngine };

File: quantsolve-server/src/engine-cpp/src/engine_runner.cpp
#include "quantsolve/engine_runner.hpp"

#include "quantsolve/error_handler.hpp"
#include "quantsolve/lexer.hpp"
#include "quantsolve/normalizer.hpp"
#include "quantsolve/parser.hpp"
#include "quantsolve/polynomial_solver.hpp"
#include "quantsolve/result_formatter.hpp"
#include "quantsolve/solver.hpp"

#include <algorithm>
#include <cctype>

namespace quantsolve {

namespace {

std::string clean_whitespace(const std::string& in) {
    std::string out;
    out.reserve(in.size());
    bool prev_space = false;
    for (unsigned char c : in) {
        if (std::isspace(c)) {
            if (!prev_space) out.push_back(' ');
            prev_space = true;
        } else {
            out.push_back(static_cast<char>(c));
            prev_space = false;
        }
    }
    if (!out.empty() && out.front() == ' ') out.erase(out.begin());
    if (!out.empty() && out.back() == ' ') out.pop_back();
    return out;
}

int tree_depth(const ASTPtr& node) {
    if (!node) return 0;
    if (node->type == NodeType::Number || node->type == NodeType::Variable) return 1;
    if (node->type == NodeType::UnaryOp) return 1 + tree_depth(node->operand);
    return 1 + std::max(tree_depth(node->left), tree_depth(node->right));
}

} // namespace

EngineResponse run_engine(const std::string& input,
                          const UserConstraints& user_constraints,
                          const SolveOptions& options) {
    try {
        if (input.empty()) {
            throw EngineError(ErrorCode::EMPTY_INPUT);
        }

        const std::string cleaned_input = clean_whitespace(input);
        if (cleaned_input.empty()) {
            throw EngineError(ErrorCode::EMPTY_INPUT);
        }

        const auto tokens = lexer(cleaned_input);
        Parser parser(tokens);
        auto [left, right] = parser.parse_equation();

        SolveOptions worker_options = options;
        if (worker_options.page <= 0) worker_options.page = 1;
        if (worker_options.page_size <= 0) worker_options.page_size = 50;

        // Try polynomial path first.
        // try_solve_polynomial NEVER throws — errors are sentinel negative values in total_found:
        //   -1 = NO_SOLUTIONS,  -2 = UNBOUNDED_SEARCH,  -3 = POLYNOMIAL_UNSUPPORTED
        // We build the error EngineResponse with raw string literals — no exceptions,
        // no RTTI, no EngineError construction — 100% cross-ABI safe.
        auto poly_result = try_solve_polynomial(left, right, user_constraints, worker_options);

        if (poly_result.has_value()) {
            if (poly_result->total_found < 0) {
                // Build error payload directly — no throw, no EngineError needed
                ErrorPayload ep;
                ep.success = false;
                if (poly_result->total_found == -1) {
                    ep.code     = "NO_SOLUTIONS";
                    ep.error    = "No whole-number solutions exist in the search domain.";
                    ep.category = "solver";
                } else if (poly_result->total_found == -2) {
                    ep.code     = "UNBOUNDED_SEARCH";
                    ep.error    = "Infinite answers detected. Please apply market limits.";
                    ep.category = "semantic";
                } else {
                    ep.code     = "POLYNOMIAL_UNSUPPORTED";
                    ep.error    = "Cannot convert equation to polynomial form.";
                    ep.category = "semantic";
                }
                EngineResponse err;
                err.success          = false;
                err.input            = cleaned_input;
                err.solutions        = {};
                err.formatted_result = FormattedResult{};
                err.warnings         = {};
                err.meta             = EngineMeta{};
                err.error_payload    = ep;
                return err;
            }

            auto& poly = *poly_result;
            FormattedResult formatted = format_results(
                poly.solutions,
                poly.variable_order,
                poly.total_found,
                poly.page,
                poly.page_size,
                poly.total_pages,
                poly.has_more
            );

            EngineResponse response;
            response.success = true;
            response.input = cleaned_input;
            response.coeffs = {};
            response.target = 0;
            response.variable_order = poly.variable_order;
            response.total_found = poly.total_found;
            response.page = poly.page;
            response.page_size = poly.page_size;
            response.total_pages = poly.total_pages;
            response.has_more = poly.has_more;
            response.solutions = poly.solutions;
            response.formatted_result = formatted;
            response.warnings = formatted.warnings;
            response.polynomial = poly.polynomial;
            response.meta = EngineMeta{
                static_cast<Int>(poly.variable_order.size()),
                static_cast<Int>(user_constraints.size()),
                tree_depth(left),
                "polynomial",
                static_cast<Int>(poly.polynomial.degree),
                poly.search_bounds,
            };

            return response;
        }


        auto norm = normalize_equation(left, right);

        std::vector<std::string> variable_order;
        variable_order.reserve(norm.coeffs.size());
        for (const auto& [k, _] : norm.coeffs) variable_order.push_back(k);
        std::sort(variable_order.begin(), variable_order.end()); // deterministic

        SolveResult solve_result = solve(norm.coeffs, norm.target, user_constraints, worker_options);

        FormattedResult formatted = format_results(
            solve_result.solutions,
            variable_order,
            solve_result.total_found,
            solve_result.page,
            solve_result.page_size,
            solve_result.total_pages,
            solve_result.has_more
        );

        EngineResponse response;
        response.success = true;
        response.input = cleaned_input;
        response.coeffs = norm.coeffs;
        response.target = norm.target;
        response.variable_order = variable_order;
        response.total_found = solve_result.total_found;
        response.page = solve_result.page;
        response.page_size = solve_result.page_size;
        response.total_pages = solve_result.total_pages;
        response.has_more = solve_result.has_more;
        response.solutions = solve_result.solutions;
        response.formatted_result = formatted;
        response.warnings = formatted.warnings;
        response.meta = EngineMeta{
            static_cast<Int>(norm.coeffs.size()),
            static_cast<Int>(user_constraints.size()),
            tree_depth(left),
            "linear",
            1,
        };

        return response;
    } catch (const std::exception& err) {
        EngineResponse response;
        response.success = false;
        response.input = input;
        response.solutions = {};
        response.formatted_result = FormattedResult{};
        response.warnings = {};
        response.meta = EngineMeta{};
        response.error_payload = handle_error(err);
        return response;
    }
}

} // namespace quantsolve
